dev/com.ibm.ws.cdi.internal/src/com/ibm/ws/cdi/internal/archive/liberty/ExtensionArchiveImpl.java

```diff
@@ -61,7 +69,22 @@ public boolean isExtClassesOnly() {
     }
 
     @Override
-    public Set<String> getSPIExtensionClasses() {
-        return spiExtensions;
+    public Set<Supplier<Object>> getSPIExtensionSuppliers() {
+        if (spiExtensions.isEmpty()) {
+            return Collections.emptySet();
+        }
+
+        Set<Supplier<Object>> result = new HashSet<>();
+        for (String className : spiExtensions) {
+            try {
+                Class<?> clazz = getClassLoader().loadClass(className);
+                Constructor<?> constructor = clazz.getConstructor();
+                Object instance = constructor.newInstance();
```

Comment:
the comment on the other method says that we are returning strings in order to delay the instantiation of the class. don't we want to do that here too? doesn't calling newInstance here defeat the point of using the supplier?

Reply:
Fair point, the class load and instantiation should possibly be inside the supplier. Will just have to check the error handling.

Reply:
On further investigation, I don't think this is necessary. This code is for creating extensions contributed via the CDIExtensionMetadata service where they're provided as `Class` objects anyway. There's no reason for this code to be doing a load by name, we could be passing through the classes to here, but even if we are loading by name we can be pretty sure it will work and work fast because the class is already loaded.

The previous method covers regular extensions as well, which are loaded by name after being found from the services file.

Reply:
No, we can't return classes because `LiteExtensionArchive` needs to be able to call a different constructor.

I'm a bit torn between just returning the extension object itself, or returning a supplier that calls the constructor. I guess it should be one or the other.